File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0, r = nums.length - 1, m = 0, originalStart = 0;
        while (l <= r) {
            m = l + (r - l)/2;
            if (isOriginal(nums, m)) {
                originalStart = m;
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        if (target >= nums[originalStart] && target <= nums[nums.length-1]) {
            l = originalStart; r = nums.length - 1;
        } else {
            l = 0; r = originalStart - 1;
        }
        if (target > nums[nums.length-1]) {
            l = 0; r = originalStart - 1;
        }
        while (l <= r) {
            m = l + (r - l)/2;
            if (nums[m] == target) return m;
            if (nums[m] < target) l = m + 1;
            else if (nums[m] > target) r = m - 1;
        }
        return -1;
    }
    private boolean isOriginal(int[] nums, int m) {
        return nums[m] <= nums[nums.length-1];
    }
}

File: Data Structures & Algorithms/target-sum/submission-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        # dp[i][s] use first i element, how many ways to sum to s
        # since s can be negative and can't be array index
        # make each dp a dict, so that the 2nd "dimension" can be negative 
        # defaultdict(int) here means for a non-existing key, 
        # default return int(), while int()=0
        dp = [defaultdict(int) for _ in range(n + 1)]
        dp[0][0] = 1

        for i in range(1, n+1):
            cur = nums[i-1]
            for s, ways in dp[i-1].items():
                dp[i][s+cur] += ways
                dp[i][s-cur] += ways 
        
        return dp[n][target]